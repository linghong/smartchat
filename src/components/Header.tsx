import { FC, MouseEvent } from 'react';
import { AiOutlineMenu, AiOutlineForm } from 'react-icons/ai';
import { useRouter } from 'next/router';

interface HeaderProps {
  setIsSidebarOpen: (isSideBarOpen: boolean) => void;
  isSidebarOpen: boolean;
  isPanelVisible: boolean;
  setIsPanelVisible: (isPanelVisible: boolean) => void;
}

const Header: FC<HeaderProps> = ({
  isSidebarOpen,
  setIsSidebarOpen,
  isPanelVisible,
  setIsPanelVisible
}) => {
  const router = useRouter();
  const pathName = router.pathname;

  const pageCollection: { [key: string]: string } = {
    '/': 'Chat With AI',
    '/finetunemodel': 'Finetune AI Model',
    '/embedragfile': 'Embed RAG File'
  };

  const onNewChat = async () => {
    // await its completion to ensure that any state changes occur after navigating, so that the sidebar won't open before the navigation completes
    await router.push('/');
    if (window.innerWidth <= 480) {
      setIsSidebarOpen(false);
    }
  };

  const toggleSidebar = (e: MouseEvent<HTMLButtonElement>) => {
    setIsSidebarOpen(!isSidebarOpen);
  };

  const toggleConfigPanel = (e: MouseEvent<HTMLButtonElement>) => {
    console.log('herr', isPanelVisible);
    setIsPanelVisible(!isPanelVisible);
  };

  return (
    <header className="w-full flex flx-col justify-between items-center bg-slate-400 p-2">
      <div className="flex w-1/3 xs:w-36 sm:w-52 md:w-48 lg:w-52 xl:w-56 items-center justify-between px-6 text-white text-md focus:bg-indigo-100">
        <button
          className={`${isSidebarOpen ? 'bg-slate-500 hover:bg-slate-700 focus:bg-stone-600' : ''}`}
          onClick={toggleSidebar}
          aria-label="Toggle Sidebar"
          role="toogle"
        >
          <AiOutlineMenu size={20} />
        </button>
        <button
          className="hover:text-lg hover:bg-stone-600 hover:font-bold"
          onClick={onNewChat}
          aria-label="New Chat"
        >
          <AiOutlineForm size={21} />
        </button>
      </div>
      <div className="flex flex-grow items-center justify-center  text-white text-md font-bold focus:bg-indigo-100 mx-auto">
        <div className="text-xl font-bold text-center text-32xl">
          <h1>{pageCollection[pathName]}</h1>
        </div>
        <button
          onClick={toggleConfigPanel}
          className="px-3 py-1 text-black rounded mb-2"
        >
          {isPanelVisible ? 'Hide' : 'Show'} Config
        </button>
      </div>
    </header>
  );
};

export default Header;
