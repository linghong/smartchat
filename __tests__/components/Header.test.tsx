import React from 'react'
import { render, screen, fireEvent } from '@testing-library/react'
import '@testing-library/jest-dom'
import { useRouter } from 'next/router'

import Header from '@/src/components/Header'

jest.mock('next/router', () => ({
  useRouter: jest.fn(),
}))

const setIsSidebarOpen = jest.fn()
const mockPush = jest.fn()

describe('Header Component', () => {
  beforeEach(() => {
    // Reset the mock before each test
    (useRouter as jest.Mock).mockReturnValue({
      pathname: '/',
      push: mockPush,
    })
    setIsSidebarOpen.mockClear()
    mockPush.mockClear()
  })

  it('should render the page title correctly', () => {
    render(<Header isSidebarOpen={true} setIsSidebarOpen={setIsSidebarOpen} />)

    // Check if the rendered header contains the correct page title
    expect(screen.getByText('Chat With AI')).toBeInTheDocument()
    expect(screen.getByRole('heading')).toHaveTextContent('Chat With AI')
  })

  it('should render different page titles based on pathname', () => {
    const paths = [
      { path: '/', title: 'Chat With AI' },
      { path: '/finetunemodel', title: 'Finetune AI Model' },
      { path: '/managemyai', title: 'Manage RAG Files' },
    ]

    paths.forEach(({ path, title }) => {
      (useRouter as jest.Mock).mockReturnValue({
        pathname: path,
        push: mockPush
      })

      render(<Header isSidebarOpen={true} setIsSidebarOpen={setIsSidebarOpen} />)

      const headings = screen.getAllByRole('heading')

      expect(headings.some(heading => heading.textContent === title)).toBe(true)
    })
  })

  describe('Mouse focusing and hovering events', () => {
    beforeEach(()=> {
      render(<Header isSidebarOpen={true} setIsSidebarOpen={setIsSidebarOpen} />)
    })

    it('should change its background color when focusing on menu button and sidebar is open', async () => {
      const menuIcon = screen.getByLabelText("Toggle Sidebar")
      expect(menuIcon?.className).toMatch(/focus/)
    })

    it('should show visual feedback when hovering over menu button and sidebar is open ', async () => {
      const menuIcon = screen.getByLabelText('Toggle Sidebar')  
      expect(menuIcon.className).toMatch(/hover/)
    })

    it('should show visual feedback when hovering over New Chat ', async () => {
      const newChatButton = screen.getByLabelText('New Chat')
      expect(newChatButton?.className).toMatch(/hover/)
    })
  })

  describe('Mouse click events on NewChat', () => {
    it('should navigate to the home page when NewChat button is clicked', async () => {
      render(<Header isSidebarOpen={true} setIsSidebarOpen={setIsSidebarOpen} />)

      fireEvent.click(screen.getByLabelText('New Chat'))      
      expect(mockPush).toHaveBeenCalledTimes(1)
      expect(mockPush).toHaveBeenCalledWith('/')
    })
    
    it('should not call setIsSidebarOpen when NewChat button is clicked on mobile and sidebar is initially open', () => {
      global.innerWidth = 480
      render(<Header isSidebarOpen={true} setIsSidebarOpen={setIsSidebarOpen} />)

      fireEvent.click(screen.getByLabelText('New Chat'))  
      // Because router.push('/') is called instead
      expect(setIsSidebarOpen).not.toHaveBeenCalled()
    })

    it('should not call setIsSidebarOpen when NewChat button is clicked on desktop and sidebar is initially open', () => {
      global.innerWidth = 1024
      render(<Header isSidebarOpen={true} setIsSidebarOpen={setIsSidebarOpen} />)

      fireEvent.click(screen.getByLabelText('New Chat'))
      expect(setIsSidebarOpen).not.toHaveBeenCalled()
    })

    it('should not call setIsSidebarOpen when NewChat button is clicked on desktop and sidebar is initially closed', () => {

      global.innerWidth = 1024
      render(<Header isSidebarOpen={false} setIsSidebarOpen={setIsSidebarOpen} />)

      fireEvent.click(screen.getByLabelText('New Chat'))
      expect(setIsSidebarOpen).not.toHaveBeenCalled()
    })

    it('should not call setIsSidebarOpen when NewChat button is clicked on mobile and sidebar is initially closed', () => {

      global.innerWidth = 480
      render(<Header isSidebarOpen={false} setIsSidebarOpen={setIsSidebarOpen} />)

      fireEvent.click(screen.getByLabelText('New Chat'))
      expect(setIsSidebarOpen).not.toHaveBeenCalled()
    }) 
  })

  describe('Mouse click events on menu icon', () => {
    it('should call setIsSidebarOpen with true when initially closed and toggle button is clicked', () => {

      render(<Header isSidebarOpen={false} setIsSidebarOpen={setIsSidebarOpen} />)

      fireEvent.click(screen.getByLabelText('Toggle Sidebar'))
      expect(setIsSidebarOpen).toHaveBeenCalledWith(true)
    })

    it('should call setIsSidebarOpen with false when initially open and toggle button is clicked', () => {

      render(<Header isSidebarOpen={true} setIsSidebarOpen={setIsSidebarOpen} />)

      fireEvent.click(screen.getByLabelText('Toggle Sidebar'))
      expect(setIsSidebarOpen).toHaveBeenCalledWith(false)
    })
  })

  it('Header component snapshot', () => {

    const { asFragment } = render(<Header isSidebarOpen={true} setIsSidebarOpen={setIsSidebarOpen} />)
    expect(asFragment()).toMatchSnapshot()
  })
})