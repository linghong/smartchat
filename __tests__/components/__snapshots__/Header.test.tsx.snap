// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header Component Header component snapshot 1`] = `
<DocumentFragment>
  <header
    class="w-full flex flx-col justify-between items-center bg-slate-400 p-2 z-50"
  >
    <div
      class="flex w-1/3 xs:w-36 sm:w-52 md:w-48 lg:w-52 xl:w-56 items-center justify-between px-6 text-white text-md focus:bg-indigo-100"
    >
      <button
        aria-label="Toggle Sidebar"
        class=" transition-colors duration-200 bg-slate-500 hover:bg-slate-700 focus:bg-stone-600"
        role="toogle"
      >
        <svg
          fill="currentColor"
          height="20"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 1024 1024"
          width="20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M904 160H120c-4.4 0-8 3.6-8 8v64c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-64c0-4.4-3.6-8-8-8zm0 624H120c-4.4 0-8 3.6-8 8v64c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-64c0-4.4-3.6-8-8-8zm0-312H120c-4.4 0-8 3.6-8 8v64c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-64c0-4.4-3.6-8-8-8z"
          />
        </svg>
      </button>
      <button
        aria-label="New Chat"
        class="hover:text-lg hover:bg-stone-600 hover:font-bold transition-colors duration-200"
      >
        <svg
          fill="currentColor"
          height="21"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 1024 1024"
          width="21"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M904 512h-56c-4.4 0-8 3.6-8 8v320H184V184h320c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H144c-17.7 0-32 14.3-32 32v736c0 17.7 14.3 32 32 32h736c17.7 0 32-14.3 32-32V520c0-4.4-3.6-8-8-8z"
          />
          <path
            d="M355.9 534.9L354 653.8c-.1 8.9 7.1 16.2 16 16.2h.4l118-2.9c2-.1 4-.9 5.4-2.3l415.9-415c3.1-3.1 3.1-8.2 0-11.3L785.4 114.3c-1.6-1.6-3.6-2.3-5.7-2.3s-4.1.8-5.7 2.3l-415.8 415a8.3 8.3 0 0 0-2.3 5.6zm63.5 23.6L779.7 199l45.2 45.1-360.5 359.7-45.7 1.1.7-46.4z"
          />
        </svg>
      </button>
    </div>
    <div
      class="flex text-white text-md font-bold focus:bg-indigo-100"
    >
      Chat With My AI Assistant
    </div>
    <div
      class="flex space-x-6"
    >
      <button
        aria-label="Show Config"
        class="flex items-center space-x-2 rounded-full hover:bg-slate-500 transition-colors duration-200"
      >
        <svg
          class="text-white"
          fill="currentColor"
          height="20"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 1024 1024"
          width="20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M942.2 486.2C847.4 286.5 704.1 186 512 186c-192.2 0-335.4 100.5-430.2 300.3a60.3 60.3 0 0 0 0 51.5C176.6 737.5 319.9 838 512 838c192.2 0 335.4-100.5 430.2-300.3 7.7-16.2 7.7-35 0-51.5zM512 766c-161.3 0-279.4-81.8-362.7-254C232.6 339.8 350.7 258 512 258c161.3 0 279.4 81.8 362.7 254C791.5 684.2 673.4 766 512 766zm-4-430c-97.2 0-176 78.8-176 176s78.8 176 176 176 176-78.8 176-176-78.8-176-176-176zm0 288c-61.9 0-112-50.1-112-112s50.1-112 112-112 112 50.1 112 112-50.1 112-112 112z"
          />
        </svg>
        <span
          class="text-white text-sm hidden sm:inline"
        >
          Show AI Assistant Config
        </span>
      </button>
    </div>
  </header>
</DocumentFragment>
`;
